rnd/autogpt_server/autogpt_server/executor/manager.py

```diff
@@ -350,76 +354,111 @@ def on_node_execution(cls, q: ExecutionQueue[NodeExecution], data: NodeExecution
         prefix = get_log_prefix(graph_data.graph_exec_id, "*")
         logger.warning(f"{prefix} Start graph execution")
 
+        finished = False
+
+        def cancel_handler():
+            while not cancel.is_set():
+                cancel.wait(1)
+            if finished:
+                return
+            cls.executor.terminate()
+            logger.info(
+                f"{prefix} Terminated graph execution {graph_data.graph_exec_id}"
+            )
+            cls._init_node_executor_pool()
+
+        cancel_thread = threading.Thread(target=cancel_handler)
+        cancel_thread.start()
+
         try:
             queue = ExecutionQueue[NodeExecution]()
             for node_exec in graph_data.start_node_execs:
                 queue.add(node_exec)
 
-            futures: dict[str, Future] = {}
+            running_executions: dict[str, AsyncResult] = {}
             while not queue.empty():
-                execution = queue.get()
+                if cancel.is_set():
+                    return
+
+                exec_data = queue.get()
 
                 # Avoid parallel execution of the same node.
-                fut = futures.get(execution.node_id)
-                if fut and not fut.done():
+                execution = running_executions.get(exec_data.node_id)
+                if execution and not execution.ready():
                     # TODO (performance improvement):
                     #   Wait for the completion of the same node execution is blocking.
                     #   To improve this we need a separate queue for each node.
                     #   Re-enqueueing the data back to the queue will disrupt the order.
-                    cls.wait_future(fut, timeout=None)
+                    execution.wait()
 
-                futures[execution.node_id] = cls.executor.submit(
-                    cls.on_node_execution, queue, execution
+                running_executions[exec_data.node_id] = cls.executor.apply_async(
+                    cls.on_node_execution,
+                    (queue, exec_data),
+                    callback=lambda _: running_executions.pop(exec_data.node_id),
                 )
 
                 # Avoid terminating graph execution when some nodes are still running.
-                while queue.empty() and futures:
-                    for node_id, future in list(futures.items()):
-                        if future.done():
-                            del futures[node_id]
-                        elif queue.empty():
-                            cls.wait_future(future)
+                while queue.empty() and running_executions:
+                    for execution in list(running_executions.values()):
+                        if cancel.is_set():
+                            return
+
+                        if not queue.empty():
+                            break  # yield to parent loop to execute new queue items
+
+                        execution.wait(3)
```

Comment:
I have tried to run a profiler and compare the flamegraph between this PR and the original one, and this seems to be the source of the CPU contention.

There seems to be something wrong here, you can try add a print statement before the wait here, then run `pytest -s test/executor/test_manager.py`.

execution.wait doesn't seem to wait at all and the CPU is spin-locking.

Reply:
There were two issues:
* The `Pool` was calling the wrong callback in some instances, causing the execution "lock" on the node (`running_executions[node_id]`) not to be popped. See [above](https://github.com/Significant-Gravitas/AutoGPT/pull/7892#discussion_r1740108876) for further details.
* The `KeyedMutex` is *deleting* locks from its `locks` store on releasing them, causing the second release to fail and the resource to remain blocked. #7936 should fix this.

Reply:
I have found another source of reduced performance. `ProcessPoolExecutor` spawns processes dynamically based on need, while `multiprocessing.Pool` spawns all of its children at once. Spawning these 10 child processes takes a few seconds.

Reply:
Reducing the number of node executors to 5 per graph executor lets it pass in CI.